### PR TITLE
April tags world

### LINKED ISSUE #31: Ground truth with AprilTags
Ground truth in the **real robot** can be done using AprilTags. There is a ROS package that implements AprilTags called [`apriltags_ros`](https://github.com/RIVeR-Lab/apriltags_ros).

The following links explain what an AprilTag is and why it's used for global localization:

- [AprilTags Visual Fiducial System](https://april.eecs.umich.edu/software/apriltag)
- [Localizing with AprilTags](http://wiki.lofarolabs.com/index.php/Localizing_with_AprilTags)
- [indoor self-localization.pdf](https://github.com/RoboticaUtnFrba/create_autonomy/files/3100756/indoor.self-localization.pdf)
- [Carrot navigation with AR Tags](https://www.youtube.com/watch?v=QkUZWRe5f4I)
- [ar_track_alvar](http://wiki.ros.org/ar_track_alvar)
- [AR Tags in Gazebo](https://www.youtube.com/watch?v=WDhIaVOUwsk) and [parts 2](https://www.youtube.com/watch?v=8aQGe18eGOw)

Note: there is no need to implement this in the Gazebo simulator because you already have a ground truth localization in the `/gts` topic.

### PR DESCRIPTION
**Pull request by serraramiro1**
Added april tag world and required functionality with [this](http://wiki.ros.org/apriltag_ros) package.
image
Added debugging nodes and configs, 
image

### CODE REVIEW

**Comment by eborghi10** [comment]
@serraramiro1 check if you can add some brightness to the tags.

**Comment by serraramiro1** [comment]
> @serraramiro1 check if you can add some brightness to the tags.

Changed layout, added lighting too.

File: ca_gazebo/launch/apriltag.launch [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<launch>
```
Comment (eborghi10) [inline ca_gazebo/launch/apriltag.launch]: Maintain the naming convention for launch files.

File: ca_gazebo/launch/apriltag.launch [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	
+	<param name="create1/amcl/initial_pose_x" value="2.15"/>
+	<param name="create1/amcl/initial_pose_y" value="-0.2"/>
+	<param name="create1/amcl/initial_pose_a" value="0"/>
+	<param name="create2/amcl/initial_pose_x" value="5"/>
+	<param name="create2/amcl/initial_pose_y" value="1"/>
+	<param name="create2/amcl/initial_pose_a" value="1.5708"/>
+
+
+	<!-- Spawn robot/s -->
```
Comment (eborghi10) [inline ca_gazebo/launch/apriltag.launch]: Are these lines necessary? (I'm comparing this launch file with others and they don't have them).

File: ca_gazebo/models/apriltag1/model.sdf [hunk 3]
```diff
@@ -0,0 +1,49 @@
+<?xml version='1.0'?>
+<sdf version='1.6'>
+  <model name='apriltag1'>
+    <link name='link'>
+      <pose frame=''>0 0 0 0 0 0</pose>
+      <inertial>
```
Comment (eborghi10) [inline ca_gazebo/models/apriltag1/model.sdf]: Is this `<inertial>` block necessary?

File: ca_gazebo/models/apriltag2/model.config [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" ?>
+<model>
+    <name>demo</name>
```
Comment (eborghi10) [inline ca_gazebo/models/apriltag2/model.config]: Use a proper name and fill the other information accordingly.

File: ca_gazebo/models/camera/model.sdf [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        <geometry>
+        <mesh>
+          <uri>model://camera/meshes/camera.stl</uri>
+          <scale>2.5 2.5 2.5</scale>
+        </mesh>
+        </geometry>
+      </visual>
+      <sensor name='my_camera' type='camera'>
+        <camera>
+          <save enabled="true">
```
Comment (eborghi10) [inline ca_gazebo/models/camera/model.sdf]: I don't recommend you to save the frames. That's not necessary.

File: ca_gazebo/models/camera/model.sdf [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            <path>/tmp/camera_save_tutorial</path>
+          </save>
+          <horizontal_fov>1.047</horizontal_fov>
```
Comment (eborghi10) [inline ca_gazebo/models/camera/model.sdf]: How much is this in degrees? Put a comment, please.

File: ca_gazebo/worlds/apriltags.world [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      <name>sun</name>
+    </include>
+    <physics type="ode">
+      <real_time_update_rate>1000.0</real_time_update_rate>
+      <max_step_size>0.001</max_step_size>
+      <real_time_factor>1</real_time_factor>
+      <ode>
+        <solver>
+          <type>quick</type>
+          <iters>150</iters>
```
Comment (eborghi10) [inline ca_gazebo/worlds/apriltags.world:33]: Compare this value with other launch files. Use the same approach.

**Review by eborghi10 — CHANGES_REQUESTED** [review]

File: ca_description/urdf/sensors/apriltag.xacro [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        <link name="april_tag_link">
+            <visual>
+                <origin xyz="0 0 0" rpy="0 0 0" />
+                <geometry>
+                    <box size="0.5 0.5 0.01"/>
+                </geometry>
+            </visual>
+        </link>
+        <joint name="april_tag_joint" type="fixed">
+            <origin xyz="0 0 ${april_tag_offset_z}" rpy="0 0 0" />
```
Comment (eborghi10) [inline ca_description/urdf/sensors/apriltag.xacro]: Pass the whole origin block as an argument.

File: ca_description/urdf/sensors/apriltag.xacro [hunk 9]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0"?>
+<robot xmlns:xacro="http://ros.org/wiki/xacro">
+    <xacro:macro name="april_tag">
+        <xacro:property name="april_tag_offset_z" value="0.25"/>
+        <link name="april_tag_link">
+            <visual>
+                <origin xyz="0 0 0" rpy="0 0 0" />
```
Comment (eborghi10) [inline ca_description/urdf/sensors/apriltag.xacro]: Remember that this line is not mandatory if the value is by default.

File: ca_description/urdf/sensors/apriltag.xacro [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+                <geometry>
+                    <box size="0.5 0.5 0.01"/>
+                </geometry>
+            </visual>
+        </link>
+        <joint name="april_tag_joint" type="fixed">
+            <origin xyz="0 0 ${april_tag_offset_z}" rpy="0 0 0" />
+            <parent link="base_link" />
+            <child link="april_tag_link" />
+            <axis xyz="1 0 0"/>
```
Comment (eborghi10) [inline ca_description/urdf/sensors/apriltag.xacro]: Is this necessary for fixed joints?

File: ca_description/urdf/sensors/test.urdf [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" ?>
```
Comment (eborghi10) [inline ca_description/urdf/sensors/test.urdf]: Is this file necessary?

File: ca_description/urdf/sensors/tmp.urdf [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" ?>
```
Comment (eborghi10) [inline ca_description/urdf/sensors/tmp.urdf]: How are you adding files to the stash? I recommend you to use:
```
git add file_name1 file_name2
git commit -m "message"
```

With this approach, you will avoid pushing code like this one.

**Review by eborghi10 — CHANGES_REQUESTED** [review]
Even more comments.

File: ca_gazebo/worlds/apriltags.world (hunk 7 above)
Comment (serraramiro1) [reply ca_gazebo/worlds/apriltags.world:33]: The values are the same than `house.world` should I use another?

File: ca_gazebo/launch/apriltag.launch (hunk 1 above)
Comment (serraramiro1) [reply ca_gazebo/launch/apriltag.launch]: Adressed

File: ca_gazebo/models/apriltag1/model.sdf (hunk 3 above)
Comment (serraramiro1) [reply ca_gazebo/models/apriltag1/model.sdf]: Adressed

File: ca_gazebo/models/apriltag2/model.config (hunk 4 above)
Comment (serraramiro1) [reply ca_gazebo/models/apriltag2/model.config]: Adressed

File: ca_gazebo/models/camera/model.sdf (hunk 5 above)
Comment (serraramiro1) [reply ca_gazebo/models/camera/model.sdf]: Adressed

File: ca_gazebo/models/camera/model.sdf (hunk 6 above)
Comment (serraramiro1) [reply ca_gazebo/models/camera/model.sdf]: Adressed

File: ca_description/urdf/sensors/apriltag.xacro (hunk 8 above)
Comment (serraramiro1) [reply ca_description/urdf/sensors/apriltag.xacro]: Adressed

File: ca_description/urdf/sensors/apriltag.xacro (hunk 9 above)
Comment (serraramiro1) [reply ca_description/urdf/sensors/apriltag.xacro]: Adressed

File: ca_gazebo/launch/apriltag.launch (hunk 2 above)
Comment (serraramiro1) [reply ca_gazebo/launch/apriltag.launch]: Adressed

File: ca_description/urdf/sensors/apriltag.xacro (hunk 10 above)
Comment (serraramiro1) [reply ca_description/urdf/sensors/apriltag.xacro]: Adressed

File: ca_description/urdf/sensors/test.urdf (hunk 11 above)
Comment (serraramiro1) [reply ca_description/urdf/sensors/test.urdf]: Adressed

File: ca_description/urdf/sensors/tmp.urdf (hunk 12 above)
Comment (serraramiro1) [reply ca_description/urdf/sensors/tmp.urdf]: Will use this from now on.

File: ca_gazebo/worlds/apriltags.world (hunk 7 above)
Comment (eborghi10) [reply ca_gazebo/worlds/apriltags.world:33]: This value can be decreased/increased depending on the complexity of the world. But it's fine :+1:

File: ca_description/urdf/sensors/apriltag.xacro [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+                    <box size="0.5 0.5 0.01"/>
+                </geometry>
+            </visual>
+        </link>
+        <joint name="april_tag_joint" type="fixed">
+            <xacro:insert_block name="origin" />
+            <parent link="base_link" />
+            <child link="april_tag_link" />
+        </joint>
+        <!---->
```
Comment (eborghi10) [inline ca_description/urdf/sensors/apriltag.xacro]: Is this necessary?

File: ca_description/urdf/sensors/apriltag.xacro [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+                        <script>
+                            <uri>model://apriltag1/materials/scripts/textures.material</uri>
+                            <uri>model://apriltag1/materials/textures/apriltag2.png</uri>
+                            <name>TexturedShape/Apriltag2</name>
+                        </script>
+                    </material>
+            </visual>
+        </gazebo>
+    </xacro:macro>
+</robot>
```
Comment (eborghi10) [inline ca_description/urdf/sensors/apriltag.xacro]: Add one empty line at the end of the files.

File: ca_description/urdf/test.urdf [hunk 15]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" ?>
```
Comment (eborghi10) [inline ca_description/urdf/test.urdf]: Remove these files please.

File: ca_gazebo/launch/create_apriltag.launch [hunk 16]
```diff
@@ -0,0 +1,27 @@
+<launch>
+	<arg name="env" default="apriltags"/>
+  	<arg name="paused" value="$(optenv PAUSED false)"/>
+	<arg name="gui" value="$(optenv GUI true)"/>
+	<arg name="debug" value="$(optenv DEBUG false)"/>
+	<env name="GAZEBO_MODEL_PATH" value="$(find ca_gazebo)/models"/>
```
Comment (eborghi10) [inline ca_gazebo/launch/create_apriltag.launch]: Don't include this here. First, check if this is necessary. If so, see [here](https://github.com/RoboticaUtnFrba/create_autonomy/blob/kinetic-devel/ca_gazebo/package.xml#L24-L28).

File: ca_gazebo/launch/create_apriltag.launch [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+	<!-- Launch Gazebo world -->
+	<include file="$(find gazebo_ros)/launch/empty_world.launch">
+		<arg name="debug" value="$(arg debug)" />
+		<arg name="gui" value="$(arg gui)" />
+		<arg name="paused" value="$(arg paused)"/>
+		<arg name="use_sim_time" value="true"/>
+		<arg name="verbose" value="true"/>
+		<arg name="world_name" value="$(find ca_gazebo)/worlds/$(arg env).world"/>
+	</include>
+	<node pkg="tf" type="static_transform_publisher" name="camera_link_broadcaster" args="2.486 -1.417 9.80 3.14159 3.14159 0 map april_camera_link 100" />
```
Comment (eborghi10) [inline ca_gazebo/launch/create_apriltag.launch]: Make a comment on why do you need this.

Comment (eborghi10) [reply ca_gazebo/launch/create_apriltag.launch]: Isn't 100 Hz too much?

File: ca_gazebo/launch/create_apriltag.launch [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+
+	<!-- Spawn robot/s -->
```
Comment (eborghi10) [inline ca_gazebo/launch/create_apriltag.launch]: I don't know if I already told you this, but why are these calls here?

File: ca_gazebo/models/camera/model.config [hunk 19]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" ?>
+<model>
+    <name>camera</name>
```
Comment (eborghi10) [inline ca_gazebo/models/camera/model.config]: Use a better name. Why are you using the camera?

File: ca_gazebo/models/camera/model.sdf [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+        <mesh>
+          <uri>model://camera/meshes/camera.stl</uri>
+          <scale>2.5 2.5 2.5</scale>
+        </mesh>
+        </geometry>
+      </visual>
+      <sensor name='my_camera' type='camera'>
+        <pose>0.2 0 0 -1.577 0 0</pose>
+        <camera>
+          <save enabled="false">
```
Comment (eborghi10) [inline ca_gazebo/models/camera/model.sdf]: Remove the tag directly.

File: ca_gazebo/models/camera/model.sdf [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+            <path>/tmp/camera_save</path>
+          </save>
+          <horizontal_fov>1.047</horizontal_fov> <!-- 188.46 degrees-->
+          <image>
+            <width>1920</width>
```
Comment (eborghi10) [inline ca_gazebo/models/camera/model.sdf]: Try to reduce the frame size.

File: ca_gazebo/models/camera/model.sdf [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+            <near>0.1</near>
+            <far>100</far>
+          </clip>
+        </camera>
+        <always_on>1</always_on>
+        <update_rate>30</update_rate>
+
+          <plugin name="april_tag_camera_controller" filename="libgazebo_ros_camera.so">
+            <alwaysOn>true</alwaysOn>
+            <updateRate>0</updateRate>
```
Comment (eborghi10) [inline ca_gazebo/models/camera/model.sdf]: Don't use 0. Use 30 or 60 Hz.

File: ca_gazebo/models/camera/model.sdf [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+            <cameraName>april_tag_camera</cameraName>
+            <imageTopicName>image_raw</imageTopicName>
```
Comment (eborghi10) [inline ca_gazebo/models/camera/model.sdf]: Use a more descriptive topic name. The idea is to understand what is this topic for only reading its name.

**Review by eborghi10 — CHANGES_REQUESTED** [review]

File: ca_description/urdf/sensors/apriltag.xacro (hunk 13 above)
Comment (serraramiro1) [reply ca_description/urdf/sensors/apriltag.xacro]: Adressed

File: ca_description/urdf/sensors/apriltag.xacro (hunk 14 above)
Comment (serraramiro1) [reply ca_description/urdf/sensors/apriltag.xacro]: Adressed

File: ca_description/urdf/test.urdf (hunk 15 above)
Comment (serraramiro1) [reply ca_description/urdf/test.urdf]: Adressed

File: ca_gazebo/launch/create_apriltag.launch (hunk 16 above)
Comment (serraramiro1) [reply ca_gazebo/launch/create_apriltag.launch]: Adressed

File: ca_gazebo/launch/create_apriltag.launch (hunk 17 above)
Comment (serraramiro1) [reply ca_gazebo/launch/create_apriltag.launch]: Set to 20Hz

File: ca_gazebo/models/camera/model.config (hunk 19 above)
Comment (serraramiro1) [reply ca_gazebo/models/camera/model.config]: Changed name to april_tags_camera

File: ca_gazebo/models/camera/model.sdf (hunk 20 above)
Comment (serraramiro1) [reply ca_gazebo/models/camera/model.sdf]: Adressed

File: ca_gazebo/models/camera/model.sdf (hunk 21 above)
Comment (serraramiro1) [reply ca_gazebo/models/camera/model.sdf]: Set to 680x480

File: ca_gazebo/models/camera/model.sdf (hunk 22 above)
Comment (serraramiro1) [reply ca_gazebo/models/camera/model.sdf]: Used 60Hz

File: ca_gazebo/models/camera/model.sdf (hunk 23 above)
Comment (serraramiro1) [reply ca_gazebo/models/camera/model.sdf]: set to april_tag_camera_image_raw

File: ca_gazebo/launch/create_apriltag.launch [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+		<arg name="paused" value="$(arg paused)"/>
+		<arg name="use_sim_time" value="true"/>
+		<arg name="verbose" value="true"/>
+		<arg name="world_name" value="$(find ca_gazebo)/worlds/$(arg env).world"/>
+	</include>
+
+	<include file="$(find ca_gazebo)/launch/publish_tag_pose.launch" if="$(arg aprildebug)">
+	</include>
+	<!--Publish the transform between the camera and the map frame, to close the cycle -->
+	<node pkg="tf" type="static_transform_publisher" name="camera_link_broadcaster" args="2.486 -1.417 9.80 3.14159 3.14159 0 map april_camera_link 20" />
```
Comment (eborghi10) [inline ca_gazebo/launch/create_apriltag.launch]: Isn't tf a new package depedency?

File: ca_gazebo/launch/create_apriltag.launch (hunk 18 above)
Comment (eborghi10) [reply ca_gazebo/launch/create_apriltag.launch]: Why don't you call [this](https://github.com/RoboticaUtnFrba/create_autonomy/blob/kinetic-devel/ca_gazebo/launch/create_house.launch#L10-L12) instead?

File: ca_gazebo/launch/publish_tag_pose.launch [hunk 25]
```diff
@@ -0,0 +1,4 @@
+<launch>
+    <!-- Launches the pose publisher node -->
+    <node pkg="ca_gazebo" type="publishtagspose.py" name="april_tags_pose_publisher"/>
```
Comment (eborghi10) [inline ca_gazebo/launch/publish_tag_pose.launch]: There is no need to create only this launch file except that you're using it in more than one place.

File: ca_gazebo/launch/create_apriltag.launch [hunk 26]
```diff
@@ -0,0 +1,44 @@
+<launch>
+	<arg name="env" default="apriltags"/>
+  	<arg name="paused" value="$(optenv PAUSED false)"/>
+	<arg name="gui" value="$(optenv GUI true)"/>
+	<arg name="debug" value="$(optenv DEBUG false)"/>
+	<arg name="aprildebug" value="$(optenv APRILDEBUG false)" doc="if true launches the pose publisher node"/>
+	<arg name="rvizcfg" value="$(optenv RVIZCFG multi_robot)"/>
```
Comment (eborghi10) [inline ca_gazebo/launch/create_apriltag.launch]: Why don't you declare this directly in `spawn_multirobot`?

File: ca_gazebo/models/april_tags_camera/model.sdf [hunk 27]
```diff
@@ -0,0 +1,48 @@
+<?xml version='1.0'?>
+<sdf version='1.6'>
+  <model name='april_tags_camera'>
+    <static>true</static>
+    <pose>-1 0 2 0 1 0</pose>
```
Comment (eborghi10) [inline ca_gazebo/models/april_tags_camera/model.sdf]: This pose must be set in the world file when you include the model. Because, if you want to include another camera, it will contain this offset already.

File: ca_gazebo/models/april_tags_camera/model.sdf [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+    <link name='april_camera_link'>
+      <visual name='visual'>
+        <geometry>
+        <mesh>
+          <uri>model://april_tags_camera/meshes/camera.stl</uri>
+          <scale>2.5 2.5 2.5</scale>
```
Comment (eborghi10) [inline ca_gazebo/models/april_tags_camera/model.sdf]: Scale the mesh in an editor and don't apply this scale.

File: ca_gazebo/models/april_tags_camera/model.sdf [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+        </mesh>
+        </geometry>
+      </visual>
+      <sensor name='my_camera' type='camera'>
+        <pose>0.2 0 0 -1.577 0 0</pose>
+        <camera>
+          <horizontal_fov>1.047</horizontal_fov> <!-- 188.46 degrees-->
```
Comment (eborghi10) [inline ca_gazebo/models/april_tags_camera/model.sdf]: Why 188.46 degrees and not 180?

File: ca_gazebo/models/apriltag1/model.config [hunk 30]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" ?>
+<model>
+    <name>apriltag1</name>
+    <version>1.0</version>
+    <sdf version="1.6">model.sdf</sdf>
+    <author>
+        <name>Ramiro Serra</name>
+        <email>serraramiro@ekumenlabs.com</email>
+    </author>
+    <description></description>
```
Comment (eborghi10) [inline ca_gazebo/models/apriltag1/model.config]: Add description.

File: ca_gazebo/models/grey_square/model.sdf [hunk 31]
```diff
@@ -0,0 +1,399 @@
+<?xml version='1.0'?>
+<sdf version='1.6'>
+  <model name='grey_square'>
```
Comment (eborghi10) [inline ca_gazebo/models/grey_square/model.sdf]: Consider calling it room/big_room or something like that.

File: ca_gazebo/models/grey_square/model.sdf [hunk 32]
```diff
@@ -0,0 +1,399 @@
...
+        <geometry>
+          <box>
+            <size>5 0.2 2.5</size>
+          </box>
+        </geometry>
+        <material>
+          <script>
+            <uri>model://grey_wall/materials/scripts</uri>
+            <uri>model://grey_wall/materials/textures</uri>
+            <name>vrc/grey_wall</name>
```
Comment (eborghi10) [inline ca_gazebo/models/grey_square/model.sdf]: Where did you get this?

File: ca_gazebo/scripts/publishtagspose.py [hunk 33]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python
```
Comment (eborghi10) [inline ca_gazebo/scripts/publishtagspose.py]: rename the file to `publish_tags_poses.py `.

File: ca_gazebo/scripts/publishtagspose.py [hunk 34]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python  
+import roslib
+import rospy
+import math
+import tf
+import geometry_msgs.msg
+import turtlesim.srv
```
Comment (eborghi10) [inline ca_gazebo/scripts/publishtagspose.py]: I guess you don't need `turtlesim.srv` for what you're doing. Check other imports too.

File: ca_gazebo/scripts/publishtagspose.py [hunk 35]
```diff
@@ -0,0 +1,35 @@
...
+
+#Publishes the tag 1 pose relative to the /map frame, to "tags_gts/pose" topic
+if __name__ == '__main__':
+    rospy.init_node('tags_tf_to_pose')
```
Comment (eborghi10) [inline ca_gazebo/scripts/publishtagspose.py]: Call the node something like `tags_tf_2_map_pose`. I think it's more descriptive. I mean, add the map word over there.

File: ca_gazebo/scripts/publishtagspose.py [hunk 36]
```diff
@@ -0,0 +1,35 @@
...
+
+    listener = tf.TransformListener()
+
+    tag_pose = rospy.Publisher('tags_gts/pose', geometry_msgs.msg.PoseStamped,queue_size=1)
```
Comment (eborghi10) [inline ca_gazebo/scripts/publishtagspose.py]: Replace `tag_pose` with `tag_pose_pub`.

Comment (eborghi10) [reply ca_gazebo/scripts/publishtagspose.py]: Leave spaces between arguments.

File: ca_gazebo/scripts/publishtagspose.py [hunk 37]
```diff
@@ -0,0 +1,35 @@
...
+
+    rate = rospy.Rate(10.0)
```
Comment (eborghi10) [inline ca_gazebo/scripts/publishtagspose.py]: Make the rate customizable.

File: ca_gazebo/scripts/publishtagspose.py [hunk 38]
```diff
@@ -0,0 +1,35 @@
...
+    pose = geometry_msgs.msg.PoseStamped()
+    while not rospy.is_shutdown():
+        try:
+            (trans,rot) = listener.lookupTransform('/map', '/tag_1', rospy.Time(0))
+            pose.pose.position.x=trans[0]
+            pose.pose.position.y=trans[1]
+            pose.pose.position.z=trans[2]
+            pose.header.frame_id="/map"
```
Comment (eborghi10) [inline ca_gazebo/scripts/publishtagspose.py]: This parameter can be set outside the while loop since it will never change. Also, set the rest of the `header` fields.

File: ca_gazebo/scripts/publishtagspose.py [hunk 39]
```diff
@@ -0,0 +1,35 @@
...
+            pose.pose.orientation.x=rot[0]
+            pose.pose.orientation.y=rot[1]
+            pose.pose.orientation.z=rot[2]
+            pose.pose.orientation.w=rot[3]
+
+            tag_pose.publish(pose)
+        except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
+            continue
```
Comment (eborghi10) [inline ca_gazebo/scripts/publishtagspose.py]: Print a warning message.

File: navigation/ca_move_base/maps/apriltags.yaml [hunk 40]
```diff
@@ -0,0 +1,6 @@
+free_thresh: 0.196
```
Comment (eborghi10) [inline navigation/ca_move_base/maps/apriltags.yaml]: Instead of uploading this you could avoid calling the node `map_server` when `LOCALIZATION=none`.

**Review by eborghi10 — CHANGES_REQUESTED** [review]

File: ca_gazebo/launch/create_apriltag.launch (hunk 24 above)
Comment (serraramiro1) [reply ca_gazebo/launch/create_apriltag.launch]: Adressed

File: ca_gazebo/launch/publish_tag_pose.launch (hunk 25 above)
Comment (serraramiro1) [reply ca_gazebo/launch/publish_tag_pose.launch]: Adressed

File: ca_gazebo/launch/create_apriltag.launch (hunk 26 above)
Comment (serraramiro1) [reply ca_gazebo/launch/create_apriltag.launch]: Adressed

File: ca_gazebo/models/april_tags_camera/model.sdf (hunk 27 above)
Comment (serraramiro1) [reply ca_gazebo/models/april_tags_camera/model.sdf]: Adressed

File: ca_gazebo/models/april_tags_camera/model.sdf (hunk 28 above)
Comment (serraramiro1) [reply ca_gazebo/models/april_tags_camera/model.sdf]: Adressed

File: ca_gazebo/models/april_tags_camera/model.sdf (hunk 29 above)
Comment (serraramiro1) [reply ca_gazebo/models/april_tags_camera/model.sdf]: My bad, it was 60 deg, adressed

File: ca_gazebo/models/apriltag1/model.config (hunk 30 above)
Comment (serraramiro1) [reply ca_gazebo/models/apriltag1/model.config]: Adressed

File: ca_gazebo/models/grey_square/model.sdf (hunk 31 above)
Comment (serraramiro1) [reply ca_gazebo/models/grey_square/model.sdf]: Adressed

File: ca_gazebo/models/grey_square/model.sdf (hunk 32 above)
Comment (serraramiro1) [reply ca_gazebo/models/grey_square/model.sdf]: I got this with the gazebo GUI

File: ca_gazebo/scripts/publishtagspose.py (hunk 33 above)
Comment (serraramiro1) [reply ca_gazebo/scripts/publishtagspose.py]: Adressed

File: ca_gazebo/scripts/publishtagspose.py (hunk 34 above)
Comment (serraramiro1) [reply ca_gazebo/scripts/publishtagspose.py]: Adressed

File: ca_gazebo/scripts/publishtagspose.py (hunk 35 above)
Comment (serraramiro1) [reply ca_gazebo/scripts/publishtagspose.py]: Adressed

File: ca_gazebo/scripts/publishtagspose.py (hunk 36 above)
Comment (serraramiro1) [reply ca_gazebo/scripts/publishtagspose.py]: Adressed

File: ca_gazebo/scripts/publishtagspose.py (hunk 37 above)
Comment (serraramiro1) [reply ca_gazebo/scripts/publishtagspose.py]: Adressed

File: ca_gazebo/scripts/publishtagspose.py (hunk 38 above)
Comment (serraramiro1) [reply ca_gazebo/scripts/publishtagspose.py]: Adressed

File: ca_gazebo/scripts/publishtagspose.py (hunk 39 above)
Comment (serraramiro1) [reply ca_gazebo/scripts/publishtagspose.py]: Adressed

File: navigation/ca_move_base/maps/apriltags.yaml (hunk 40 above)
Comment (serraramiro1) [reply navigation/ca_move_base/maps/apriltags.yaml]: Deleted `apriltags.yaml`

**Comment by serraramiro1** [comment]
Ready for review @eborghi10 @ernestmc

File: ca_gazebo/launch/create_apriltag.launch [hunk 41]
```diff
@@ -1,44 +1,33 @@
 <launch>
 	<arg name="env" default="apriltags"/>
```
Comment (eborghi10) [inline ca_gazebo/launch/create_apriltag.launch]: There is no need to create this argument since you're not going to modify it.

File: ca_gazebo/launch/create_apriltag.launch [hunk 42]
```diff
@@ -1,44 +1,33 @@
...
+	<arg name="aprildebug" default="$(optenv APRILDEBUG false)" doc="if true launches the pose publisher node"/>
 
 	<!--set the position where the robot will be spawned-->
 	<param name="create1/amcl/initial_pose_x" value="2.25"/>
   	<param name="create1/amcl/initial_pose_y" value="-1.5"/>
   	<param name="create1/amcl/initial_pose_a" value="0"/>
 	<!-- Launch Gazebo world -->
-	<include file="$(find gazebo_ros)/launch/empty_world.launch">
+
+	<node pkg="ca_gazebo" type="publish_tags_poses.py" name="april_tags_pose_publisher" if="$(arg aprildebug)">
```
Comment (eborghi10) [inline ca_gazebo/launch/create_apriltag.launch]: Move the script to `ca_tools`.

File: ca_gazebo/launch/create_empty_world.launch [hunk 43]
```diff
@@ -1,8 +1,8 @@
 <launch>
 	<arg name="env" default="empty"/>
-  	<arg name="paused" value="$(optenv PAUSED false)"/>
-	<arg name="gui" value="$(optenv GUI true)"/>
-	<arg name="debug" value="$(optenv DEBUG false)"/>
+  	<arg name="paused" default="$(optenv PAUSED false)"/>
```
Comment (eborghi10) [inline ca_gazebo/launch/create_empty_world.launch:3]: Good catch!

File: ca_gazebo/worlds/apriltags.world [hunk 44]
```diff
@@ -45,14 +45,14 @@
...
       <uri>model://april_tags_camera</uri>
-      <pose>2.486 -1.417 10.0 0 1.57079 0</pose>
+      <pose>1.968 -0.84 10.0 0 1.57079 0</pose>
       <name>april_tags_camera</name>
     </include>
     <include>
-      <uri>model://grey_square</uri>
+      <uri>model://big_room</uri>
       <pose>-2.093255 -5.88 -0.3 0 0 0</pose>
       <scale>2 2 0</scale>
```
Comment (eborghi10) [inline ca_gazebo/worlds/apriltags.world]: Why are you scaling the model?

**Review by eborghi10 — CHANGES_REQUESTED** [review]

File: ca_gazebo/launch/create_apriltag.launch (hunk 41 above)
Comment (serraramiro1) [reply ca_gazebo/launch/create_apriltag.launch]: Adressed

File: ca_gazebo/launch/create_apriltag.launch (hunk 42 above)
Comment (serraramiro1) [reply ca_gazebo/launch/create_apriltag.launch]: Adressed

File: ca_gazebo/worlds/apriltags.world (hunk 44 above)
Comment (serraramiro1) [reply ca_gazebo/worlds/apriltags.world]: Removed tag

File: ca_gazebo/package.xml [hunk 45]
```diff
@@ -13,7 +13,11 @@
   <build_depend>gazebo_plugins</build_depend>
   <build_depend>roscpp</build_depend>
   <build_depend>ca_msgs</build_depend>
+  <build_depend>apriltag_ros</build_depend>
+  <build_depend>tf</build_depend>
 
+  <run_depend>apriltag_ros</run_depend>
```
Comment (ernestmc) [inline ca_gazebo/package.xml:19]: If a dependency is both a run_depend and build_depend, the it is convenient to just use `<depend>`. See [here](http://wiki.ros.org/catkin/package.xml#Dependencies)

**Review by ernestmc — APPROVED** [review]
lgtm
